clarity_ext/domain/udf.py

```diff
@@ -1,128 +1,215 @@
+        # It should be in a list, since those mapped by pyname will potentially be more
+        # than one:
+        udf_info = UdfInfo(key, value)
+        self.values.add(udf_info)
+        self.raw_map[key] = [udf_info]
+
+        # Then, we also fetch the py name, and add that to the raw map too
+        py_name = self._automap_name(key)
+        self.raw_map.setdefault(py_name, list())
+        self.raw_map[py_name].append(udf_info)
+        self.py_names.add(py_name)
+
+        # Post: The raw_map will contain a new key that corresponds to the original
+        # UDF. It will also contain a mapping from a python name to that exact same
+        # value, so fetching by either name will lead to the same results.
+
+    def __getitem__(self, key):
+        return self.unwrap(key)
+
+    def __setitem__(self, key, value):
+        self.unwrap(key).value = value
+
+    def unwrap(self, key):
         """
-        Given a dictionary of UDFs, returns a new dictionary that uses Python naming conventions instead.
+        First tries to fetch by the original key, raises an exception if there
+        are more than one possible values for the key.
 
-        Also returns a dictionary mapping from the original UDF name back to the attribute name, which can
-        be used to synchronize the two if needed.
+        Raises a KeyError if the key is not available in the UDF map
         """
-        attributes = dict()
-        # A dictionary that matches from Python attributes back to the api
-        # resource UDFs.
-        udfs_to_attributes = dict()
-        for key, value in original_udfs.items():
-            attrib_name = cls.automap_name(key)
-            attributes[attrib_name] = value
-            udfs_to_attributes[key] = attrib_name
-        return attributes, udfs_to_attributes
-
-    @classmethod
-    def automap_name(cls, original_udf_name):
+        udf_info = self.raw_map[key]
+        if len(udf_info) > 1:
+            raise UdfMappingNotUniqueException(key)
+        return udf_info[0]
+
+    def create_from_dict(self, udf_dict):
+        for key, value in udf_dict.items():
+            self.add(key, value)
+
+    def usage(self):
+        """Returns a string showing which UDFs are available, using Python names"""
+        return ", ".join(self.py_names)
+
+    def enumerate_updated(self):
+        return (value for value in self.values if value.is_dirty())
+
+    def __contains__(self, item):
+        return item in self.raw_map
+
+    @staticmethod
+    def _automap_name(original_udf_name):
         """
         Maps a UDF name from Clarity to one that matches Python naming conventions
 
-        Example: 'Fragment Lower (bp)' => 'fragment_lower_bp'
+        The naming scheme may cause clashes, which would need to be resolved by the
+        caller (e.g. by ignoring UDFs that clash, or by throwing an exception)
+
+        Examples:
+          'Fragment Lower (bp)' => 'udf_fragment_lower_bp'
+          '% Total' => 'udf_total'
```

Comment:
Add a todo in jira on handling this? What about clashes against "Total". Would it make sense to map it to udf_percent_total?